File: src/main/golang/leetcode/editor/cn/[241]Different Ways to Add Parentheses.go
package lc

import "strconv"

//leetcode submit region begin(Prohibit modification and deletion)
func diffWaysToCompute(expression string) []int {
	var ans []int
	n := len(expression)
	for i := 0; i < n; i++ {
		if expression[i] == '+' || expression[i] == '-' || expression[i] == '*' {
			left := diffWaysToCompute(expression[:i])
			right := diffWaysToCompute(expression[i+1:])
			for _, s0 := range left {
				for _, s1 := range right {
					if expression[i] == '+' {
						ans = append(ans, s0+s1)
					} else if expression[i] == '-' {
						ans = append(ans, s0-s1)
					} else if expression[i] == '*' {
						ans = append(ans, s0*s1)
					}
				}
			}
		}
	}
	if len(ans) == 0 {
		num, _ := strconv.Atoi(expression)
		ans = append(ans, num)
	}
	return ans
}

//leetcode submit region end(Prohibit modification and deletion)

File: src/main/golang/leetcode/editor/cn/[241]Different Ways to Add Parentheses.py
from functools import lru_cache
from typing import List


# leetcode submit region begin(Prohibit modification and deletion)
class Solution:
    def diffWaysToCompute(self, expression: str) -> List[int]:
        @lru_cache(None)
        def dfs(s):
            if s.isdigit():
                return [int(s)]
            ans = []
            for i, c in enumerate(s):
                if c in "+-*":
                    left = dfs(s[:i])
                    right = dfs(s[i + 1:])
                    for l in left:
                        for r in right:
                            ans.append(eval(f"{l}{c}{r}"))
            return ans
        return dfs(expression)
# leetcode submit region end(Prohibit modification and deletion)
